data/random-teams.ts

```diff
@@ -1774,17 +1792,14 @@ export class RandomTeams {
 		let numMaxLevelPokemon = 0;
 
 		const pokemonList = isDoubles ? Object.keys(this.randomDoublesSets) : Object.keys(this.randomSets);
-		const [pokemonPool, baseSpeciesPool] = this.getPokemonPool(type, pokemon, isMonotype, pokemonList);
+		const [pokemonPool, shuffledBaseSpecies] = this.getPokemonPool(type, pokemon, isMonotype, pokemonList);
 
 		let leadsRemaining = this.format.gameType === 'doubles' ? 2 : 1;
-		while (baseSpeciesPool.length && pokemon.length < this.maxTeamSize) {
-			const baseSpecies = this.sampleNoReplace(baseSpeciesPool);
-			const currentSpeciesPool: Species[] = [];
-			for (const poke of pokemonPool) {
-				const species = this.dex.species.get(poke);
-				if (species.baseSpecies === baseSpecies) currentSpeciesPool.push(species);
-			}
-			let species = this.sample(currentSpeciesPool);
+
+		while (shuffledBaseSpecies.length && pokemon.length < this.maxTeamSize) {
+			// repeated popping from weighted shuffle is equivalent to repeated weighted sampling without replacement
+			const baseSpecies = shuffledBaseSpecies.pop()!.baseSpecies;
+			let species = this.dex.species.get(this.sample(pokemonPool[baseSpecies].speciesArray));
```

Comment:
(Obviously there are some other fixes required; these suggestions are not meant to be exhaustive.)
```suggestion
			let species = this.dex.species.get(this.sample(pokemonPool[baseSpecies]));
```